toolchain/driver/driver.cpp

```diff
@@ -400,7 +400,7 @@ auto Driver::Compile(const CompileOptions& options) -> bool {
   CARBON_VLOG() << "*** SourceBuffer::CreateFromFile done ***\n";
   // Require flushing the consumer before the source buffer is destroyed,
   // because diagnostics may reference the buffer.
-  auto flush = llvm::make_scope_exit([&]() { consumer->Flush(); });
+  auto flush_for_source = llvm::make_scope_exit([&]() { consumer->Flush(); });
```

Comment:
Since this doesn't work as intended, maybe we should just explicitly flush before each return?

By my count, that'd be lines 407, 424, 439, and 461 (or maybe 463 and 467 if you want to be consistent). Either way, removing the 476 flush.

That versus a scope exit on 403, 414, 430, and 447 plus the 476 explicit flush.

Just observing as I scrutinize flushes, the flushes on 419, 434, and 451 actually seem like they should maybe be removed, as a flawed translation when the prior "dump" setup was split (i.e., dumping these shouldn't force diagnostics to print early). I'd be fine changing that in its own PR though.

Reply:
Sure, we can do that. I have a lingering concern that any of these approaches seems too error-prone. I think it'll be easy to forget to add a flush, and the bugs you get as a consequence are subtle. Maybe we should also make the consumer CHECK-fail from its destructor if it's not been flushed, rather than auto-flushing?

I think it makes sense to flush before producing dump output: in the common case of running an action from the command line, I think I'd want to see error / warning messages before the dump output, rather than afterwards. It is strange that enabling dump output will produce diagnostics in a different order, but I guess we can't do much about that unless we want to buffer all the dump output and print it after we finish the action.

Reply:
Let's decide what to do with the flush-before-dump cases in #3128.

Reply:
> Sure, we can do that. I have a lingering concern that any of these approaches seems too error-prone. I think it'll be easy to forget to add a flush, and the bugs you get as a consequence are subtle. Maybe we should also make the consumer CHECK-fail from its destructor if it's not been flushed, rather than auto-flushing?

SGTM

> I think it makes sense to flush before producing dump output: in the common case of running an action from the command line, I think I'd want to see error / warning messages before the dump output, rather than afterwards. It is strange that enabling dump output will produce diagnostics in a different order, but I guess we can't do much about that unless we want to buffer all the dump output and print it after we finish the action.

Okay... I mean, pragmatically I don't think the choice here matters much since end users probably won't notice.